lib/std/sysrand.nim

```diff
@@ -146,29 +165,40 @@ elif defined(linux):
 elif defined(openbsd):
   proc getentropy(p: pointer, size: cint): cint {.importc: "getentropy", header: "<unistd.h>".}
 
-  proc randomBytes(p: pointer, size: Natural): int =
-    while result < size:
-      let readBytes = getentropy(p, cint(size - result))
-      processReadBytes(readBytes, p, result)
-
   proc urandom*(p: var openArray[byte]): int =
     let size = p.len
     if size > 0:
-      result = randomBytes(addr p[0], size)
+      let
+        chunks = (size - 1) div batchSize
+        left = size - chunks * batchSize
+      var base = 0
+      for i in 0 ..< chunks:
+        let readBytes = getentropy(addr p[base], cint(batchSize))
+        if readBytes < 0:
+          return readBytes
+        inc(base, batchSize)
+
+      result = getentropy(addr p[base], cint(left))
 
 elif defined(freebsd):
-  type ssize_t = int
-  proc getrandom(p: pointer, size: csize_t, flags: cuint): ssize_t {.importc: "getrandom", header: "<sys/random.h>".}
+  type cssize_t {.importc: "ssize_t", header: "<sys/types.h>".} = int
 
-  proc randomBytes(p: pointer, size: int): int =
-    while result < size:
-      let readBytes = getrandom(p, csize_t(size - result), 0)
-      processReadBytes(readBytes, p, result)
+  proc getrandom(p: pointer, size: csize_t, flags: cuint): cssize_t {.importc: "getrandom", header: "<sys/random.h>".}
 
   proc urandom*(p: var openArray[byte]): int =
     let size = p.len
     if size > 0:
-      result = randomBytes(addr p[0], size)
+      let
```

Comment:
optional: can you abstract this pattern into a template? it's used twice (and in future work it could be exposed as an API `byChunks`)